File: app/api/routers/user_anime.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.api.deps_auth import get_current_user
from app.models.user import User
from app.schemas.user_anime import UserAnimeCreate, UserAnimeUpdate, UserAnimeResponse
from app.services import user_anime_service
from app.services.user_anime_service import UserAnimeAlreadyExistsError, UserAnimeNotFoundError

router = APIRouter(prefix="/list", tags=["list"])


@router.post("/", response_model=UserAnimeResponse)
def add_anime(
    data: UserAnimeCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return user_anime_service.add_anime(current_user.id, data, db)
    except UserAnimeAlreadyExistsError:
        raise HTTPException(status_code=400, detail="Anime already in list.")


@router.get("/", response_model=list[UserAnimeResponse])
def list_anime(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return user_anime_service.list_anime(current_user.id, db)


@router.patch("/{anime_id}", response_model=UserAnimeResponse)
def update_anime(
    anime_id: int,
    data: UserAnimeUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return user_anime_service.update_anime(current_user.id, anime_id, data, db)
    except UserAnimeNotFoundError:
        raise HTTPException(status_code=404, detail="Anime not found in list.")


@router.delete("/{anime_id}", status_code=204)
def remove_anime(
    anime_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        user_anime_service.remove_anime(current_user.id, anime_id, db)
    except UserAnimeNotFoundError:
        raise HTTPException(status_code=404, detail="Anime not found in list.")

File: backend-rust/src/routes/user_anime.rs
use axum::{
    extract::{Path, State},
    http::{header::AUTHORIZATION, HeaderMap, StatusCode},
    routing::{get, patch},
    Json, Router,
};

use crate::{
    app_state::AppState,
    errors::AppError,
    schemas::{
        auth::UserResponse,
        user_anime::{
            UserAnimeCreate, UserAnimeListItemResponse, UserAnimeResponse, UserAnimeUpdate,
        },
    },
    services::{auth_service, user_anime_service},
};

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/list", get(list_anime).post(add_anime))
        .route("/list/", get(list_anime).post(add_anime))
        .route("/list/{anime_id}", patch(update_anime).delete(remove_anime))
}

async fn add_anime(
    State(state): State<AppState>,
    headers: HeaderMap,
    Json(payload): Json<UserAnimeCreate>,
) -> Result<Json<UserAnimeResponse>, AppError> {
    let current_user = require_authenticated_user(&state, &headers).await?;

    let entry = user_anime_service::add_anime(&state.db, current_user.id, payload).await?;

    Ok(Json(entry))
}

async fn list_anime(
    State(state): State<AppState>,
    headers: HeaderMap,
) -> Result<Json<Vec<UserAnimeListItemResponse>>, AppError> {
    let current_user = require_authenticated_user(&state, &headers).await?;

    let list = user_anime_service::list_anime(&state.db, current_user.id).await?;

    Ok(Json(list))
}

async fn update_anime(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(anime_id): Path<i32>,
    Json(payload): Json<UserAnimeUpdate>,
) -> Result<Json<UserAnimeResponse>, AppError> {
    let current_user = require_authenticated_user(&state, &headers).await?;

    let entry =
        user_anime_service::update_anime(&state.db, current_user.id, anime_id, payload).await?;

    Ok(Json(entry))
}

async fn remove_anime(
    State(state): State<AppState>,
    headers: HeaderMap,
    Path(anime_id): Path<i32>,
) -> Result<StatusCode, AppError> {
    let current_user = require_authenticated_user(&state, &headers).await?;

    user_anime_service::remove_anime(&state.db, current_user.id, anime_id).await?;

    Ok(StatusCode::NO_CONTENT)
}

async fn require_authenticated_user(
    state: &AppState,
    headers: &HeaderMap,
) -> Result<UserResponse, AppError> {
    let authorization = headers.get(AUTHORIZATION);
    auth_service::current_user(&state.db, &state.config, authorization).await
}
